core/drivers/clk/clk.c

```diff
@@ -202,7 +202,21 @@ void clk_disable(struct clk *clk)
 
 unsigned long clk_get_rate(struct clk *clk)
 {
-	return clk->rate;
+	unsigned long rate = clk->rate;
+
+	if (clk->ops->get_rate) {
+		unsigned long parent_rate =  0UL;
+
+		if (clk->parent)
+			parent_rate = clk_get_rate(clk->parent);
+
+		return clk->ops->get_rate(clk, parent_rate);
+	}
+
+	if (clk->parent)
+		rate = clk_get_rate(clk->parent);
```

Comment:
Why? In that case we return the cached value?

Reply:
Looking into the code, `clk_compute_rate_no_lock()` already assumes clocks that have no `.get_rate` operation handler assume their rate is the rate of their parent, and rate is 0 if they have no parent.
This goes is favor of what you propose here.

That said, this commit claims to address cases where a parent clock could be change without its child/children clock(s) be notified. In that way, I think a more generic case change should be:
```patch
 static void clk_compute_rate_no_lock(struct clk *clk)
 {
 	unsigned long parent_rate = 0;
 
 	if (clk->parent)
-		parent_rate = clk->parent->rate;
+		parent_rate = clk_compute_rate_no_lock(clk->parent);
 
 	if (clk->ops->get_rate)
 		clk->rate = clk->ops->get_rate(clk, parent_rate);
	else
 		clk->rate = parent_rate;
 }

unsigned long clk_get_rate(struct clk *clk)
{
+	clk_compute_rate_no_lock(clk);
	return clk->rate;
}
```
Together with:
```patch
 /**
  * clk_get_rate() - Get clock rate
  *
  * @clk: Clock for which the rate is needed
  * Return the clock rate in Hz
+ *
+ * This function take care of clocks for which the parent clock rate
+ * changed without notifying the child clock. It is recommended to
+ * use clk_get_rate() instead of read straight @clk->rate.
  */
 unsigned long clk_get_rate(struct clk *clk);
```

Reply:
the purpose of clk_compute_rate_no_lock() is to update the cached clock rate (clk->rate) of ALL clock in clock framework when one clock is initialized or modified:

- when register in clk_register()
- when parent selection change in clk_set_parent_no_lock()
- when clock change in clk_set_rate_no_lock()

This cache allow to avoid to read registers/ compute frequency for all clock tree for each get_rate.

when you said "a parent clock could be change without its child/children clock(s) be notified."
that means someting is wrong in cached clock rate (clk->rate) 

so all the framework must change to
- correctly update the clk->rate value when the parent clock change
- OR completely remove the cached values clk->rate => always request ops .get_rate